### PR TITLE
Deprecate ByteVector

### PR DESCRIPTION
**Pull request by andres-erbsen**
On top of https://github.com/coq/coq/pull/17732

- [x] Added **changelog**.

Trial removal PR at https://github.com/coq/coq/pull/17129; Bignums will need to be adapted before actually removing ByteVector.

### CODE REVIEW

**Comment by proux01** [comment]
Just a detail: if this is just a test to run the CI and you are not actually expecting reviews yet, please open the PR in draft mode. That will avoid spaming multiple maintainers.

**Comment by andres-erbsen** [comment]
Thanks. I would like this PR with the deprecation to be reviewed, the one for removal is still draft as Bignums will need superficial changes. Diff link to the commit not included in the base PR: https://github.com/coq/coq/pull/18022/commits/6e0d15d2f21d97286434e16f40f2a26d33426e1f

**Review by proux01 — APPROVED** [review]
Let's first merge #17732 and I'll merge this one after if there are no other comments.

**Comment by proux01** [comment]
@coqbot merge now

**Comment by liyishuai** [comment]
(as the author of #8365) What motivates this deprecation? Are vector and lists the same thing?

The deprecation and removal here won't break too much, as [coq-http2](https://github.com/liyishuai/coq-http2) is out of maintenance (and beyond Coq CI). I just wonder if deprecation implies the code being duplicated/misleading/malicious, or simply unused.

**Comment by JasonGross** [comment]
My understanding is that `Vector.t` is discouraged except in cases where the length is globally fixed and there is no code that needs to deal with different lengths, mostly because dependent types are a pain in practice and when dealing with lists, it's generally better to separate the proofs of "preserves length" from the data manipulation.  My guess is that this is at least partially behind the deprecation, which would make the deprecation something like "unused and not an example of data structure usage to be encouraged"?  But I'm just guessing here.

**Comment by liyishuai** [comment]
Discouraging a data structure regardless of context sounds weird. Such documentations should belong to blog posts or Coqdocs, but not standard library warnings.

E.g., for fixed-length network packets, the recipient only deals with two cases "packet received" and "packet dropped", no "received several bytes". Dependent type allows cutting the packets cleanly at specific points (using [`splitat`](https://coq.inria.fr/distrib/current/stdlib/Coq.Vectors.VectorDef.html#splitat)), while lists require length checkers everywhere.

**Comment by proux01** [comment]
@JasonGross indeed, there were already a few discussions on Zulip, see also https://github.com/coq/coq/pull/18032

@liyishuai note that nobody is claiming that pure lists are a good replacement for `Vector.t`, rather dependent pairs with list and a proof about its length. Then if the proofs are computable, they all reduce from `(length l == n) = true` to `true = true` and canonicity is retained, as done in `tuple.v` in MathComp.

**Pull request merged**
